### PR TITLE
Tests

### PR DESCRIPTION
**Pull request by PatWie**
I started the refactoring of the tests as the previous tests were a bit hacky. Hope this is much easier to maintain. Still not happy with the `setUp` and `teastDown` duplicated code.

This addresses #153 raised in #154. 

- [x] cached MNIST dataset
- [x] testing layers (several examples)
- [x] add dummy images for more complex networks (ideally one should run build ResNet on FakeData)
- [x] remove duplicate codelines

### CODE REVIEW

**Comment by PatWie** [comment]
There are interesting findings:
- matplotlib is not in requirements.txt. I am not sure if you want to add this.
- the ResNet script relies on "import caffe". This is only for reading the mean-file, right? Do you mind to use a pickle or numpy msgpack file instead?

**Comment by ppwwyyxx** [comment]
+ matplotlib is in opt-requirements, as it's not necessary for most functionailities
+ does it? I think it only imports caffe_pb2.py, which doesn't require caffe to be installed.

**Comment by PatWie** [comment]
@ppwwyyxx Can you have a look at some point. It seems that the ilsvrc-class reads from a different file the paths of the imagenet data. I am trying to fake some image-files an update the "train.txt" and "val.txt".

However, I get
https://travis-ci.org/ppwwyyxx/tensorpack/jobs/215508412#L356

```
 File "/home/travis/virtualenv/python2.7.12/lib/python2.7/site-packages/tensorpack/dataflow/dataset/ilsvrc.py", line 184, in get_data
              assert im is not None, fname
          AssertionError: ilsvrc/train/n02481823/n02481823_2939.JPEG
```

This is surprising, as there is not such a line 'ilsvrc/train/n02481823/n02481823_2939.JPEG' in train.txt

**Comment by ppwwyyxx** [comment]
The default dataset directory for `ILSVRCMeta` is `ilsvrc_metadata`, not `ilsvrc`.
But you can pass `meta_dir` argument to `ILSVRC()` as well. It is the full path to the directory.

File: tests/test_resnet.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+import tarfile
+import errno
+from six.moves import urllib
+import sys
+
+
+CAFFE_ILSVRC12_URL = "http://dl.caffe.berkeleyvision.org/caffe_ilsvrc12.tar.gz"
+
+
+# this is duplicated code, but this file should be self-contained
```
Comment (ppwwyyxx) [inline tests/test_resnet.py]: why not import stuff from tensorpack? (as another form of testing)

**Comment by ppwwyyxx** [comment]
Btw, resnet example now uses NCHW, which is not supported by TensorFlow on CPU. You might encounter this problem later..

File: tests/test_resnet.py (hunk 1 above)
Comment (PatWie) [inline tests/test_resnet.py]: Sure. I currently thinking about using FakeData by a flag in the ResNet example, which simplifies this testing.

**Comment by PatWie** [comment]
I currently thinking about using FakeData by a flag in the ResNet example, which would simplify this testing.

File: tests/test_resnet.py (hunk 1 above)
Comment (PatWie) [inline tests/test_resnet.py]: the latest version uses the import

**Comment by PatWie** [comment]
After several edits, I was able to convince travis-ci to pass this pull-request. I changed some lines in the resnet example:
- add argument for data_format (for training)
- add argument for faking data

File: tests/test_resnet.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+import os
+import shutil
+import tarfile
+from tensorpack.utils.fs import download, mkdir_p
+
+
+CAFFE_ILSVRC12_URL = "http://dl.caffe.berkeleyvision.org/caffe_ilsvrc12.tar.gz"
+
+
+def fake_ilsvrc12():
```
Comment (ppwwyyxx) [inline tests/test_resnet.py]: Is this still needed now that you have --fake?

**Comment by ppwwyyxx** [comment]
Removed fake_ilsvrc12, doesn't seem necessary? 
Thanks very much!

**Pull request merged**